UnityProject/Assets/Scripts/Systems/StatusesAndEffects/Implementations/Convulsing.cs

```diff
@@ -0,0 +1,120 @@
+﻿using System;
+using Core.Chat;
+using StatusesAndEffects.Interfaces;
+using UnityEngine;
+using Random = UnityEngine.Random;
+
+
+namespace StatusesAndEffects.ScriptableObjects
+{
+	[CreateAssetMenu(fileName = "Convulsing", menuName = "ScriptableObjects/StatusEffects/Convulsing")]
+	public class Convulsing: StatusEffect, IStackableStatus, IExpirableStatus
+	{
+
+		private static readonly int maxStacks = 5;
+		private static readonly Vector2 triggerTimeRange = new(5f, 10f);
+		private static readonly int baseProcChance = 20;
+		public float duration = 150f;
+		public int initialStacks = 1;
+
+		public event Action<StatusEffect> Expired;
+		public float Duration => duration;
+		public int Stacks { get; private set; }
+		private DateTime DeathTime { get; set; }
+
+		private int actualProcChance;
+		private DateTime nextTriggerTime;
+		private float Timer => Random.Range(triggerTimeRange.x, triggerTimeRange.y);
+
+		public override void OnAdded()
+		{
+			Stacks = initialStacks;
+			actualProcChance = baseProcChance;
+			nextTriggerTime = DateTime.Now.AddSeconds(Timer);
+			DeathTime = DateTime.Now.AddSeconds(duration);
+			UpdateManager.Add(Trigger, 1f);
+			UpdateManager.Add(CheckExpiration, 1f);
+		}
+
+		public override void OnRemoved()
+		{
+			UpdateManager.Remove(CallbackType.PERIODIC_UPDATE, Trigger);
+			UpdateManager.Remove(CallbackType.PERIODIC_UPDATE, CheckExpiration);
+		}
+
+		public void AddStack(int amount)
+		{
+			Stacks = Mathf.Clamp(Stacks + amount, 0, maxStacks);
+			actualProcChance = baseProcChance * Stacks;
+		}
+
+		public void RemoveStack(int amount)
+		{
+			Stacks -= amount;
+			actualProcChance = baseProcChance * Stacks;
+		}
+
+
+		private void DropItemFromHands(NamedSlot namedSlot, PlayerScript player)
+		{
+			foreach (var slot in player.DynamicItemStorage.GetNamedItemSlots(namedSlot))
+			{
+				var item = slot.Item;
+				if (item != null)
+				{
+					Chat.AddActionMsgToChat(
+						target,
+						$"You convulse violently, and you drop {item.gameObject.ExpensiveName()}!",
+						$"{target.ExpensiveName()} convulses, and they drop their {item.gameObject.ExpensiveName()}!");
+				}
+				else
+				{
+					Chat.AddActionMsgToChat(
+						target,
+						$"You convulse violently!",
+						$"{target.ExpensiveName()} convulses!");
+				}
+				Inventory.ServerDrop(slot);
+			}
+		}
+
+
+		public override void DoEffect()
+		{
+			if (target.TryGetComponent<PlayerScript>(out var player) == false)
+			{
+				return;
+			}
+
+			if (DMMath.Prob(actualProcChance))
+			{
+				DropItemFromHands(NamedSlot.leftHand, player);
```

Comment:
Where does this come from?

Reply:
The DMMath operation?

Reply:
it is the method to test chances we are supposed to use. You give it a percent and it returns a bool

Reply:
NamedSlot

Reply:
oh, it's enum don't worry I'm just dumb